File: Python/_031_Next_Permutation/solution.py
class Solution:
    def nextPermutation(self, nums):
        """
        :type nums: List[int]
        :rtype: void Do not return anything, modify nums in-place instead.
        """
        consider = len(nums) - 2
        while consider >= 0:
            if nums[consider] < nums[consider + 1]: 
                break
            consider -= 1
        for j in range(consider + 1, (len(nums) + consider + 1) // 2):
            nums[j], nums[len(nums) + consider - j] = nums[len(nums) + consider - j], nums[j]
        if consider == -1: return
        for j in range(consider + 1, len(nums)):
            if nums[j] > nums[consider]:
                nums[j], nums[consider] = nums[consider], nums[j]
                break

File: Java/_031_Next_Permutation/Solution.java
class Solution {
    public void nextPermutation(int[] nums) {
        int i;
        for (i = nums.length - 2; i >= 0; i--) {
            if (nums[i] < nums[i + 1]) 
                break;
        }    
        for (int j = i + 1; j < (nums.length + i + 1) / 2; j++) {
            int temp = nums[j];
            nums[j] = nums[nums.length + i - j];
            nums[nums.length + i - j] = temp;
        }
        if (i == -1) return;
        for (int j = i + 1; j < nums.length; j++)
            if (nums[j] > nums[i]) {
                int temp = nums[j];
                nums[j] = nums[i];
                nums[i] = temp;
                break;
            }
    }
}
